singletons-th/src/Data/Singletons/TH.hs

```diff
@@ -71,8 +71,54 @@ import Language.Haskell.TH.Desugar
 -- is identical. This may be useful if the type-checker requires knowledge of which
 -- constructor is used to satisfy equality or type-class constraints, but where
 -- each constructor is treated the same.
+--
+-- Here is a simple example to illustrate where 'cases' can be useful. Suppose
+-- you use @singletons-th@ to single this code:
+--
+-- @
+-- $('singletons' [d|
+--   foo :: Bool -> ()
+--   foo True = ()
+--   foo False = ()
+--   |])
+-- @
+--
+-- And that you want to write a function of this type:
+--
+-- @
+-- bar :: SBool b -> STuple0 (Foo b)
+-- @
+--
+-- How would you do this? You might be tempted to write the following:
+--
+-- @
+-- bar _ = STuple0
+-- @
+--
+-- However, this won't typecheck, as Foo b won't reduce to @'()@ unless GHC
+-- knows @b@ is either 'True' or 'False'. In order to convince GHC of this, you
+-- must explicitly match on each of the data constructors of @SBool@:
+--
+-- @
+-- bar :: SBool b -> STuple0 (Foo b)
+-- bar b = case b of
+--   STrue  -> STuple0
+--   SFalse -> STuple0
+-- @
+--
+-- This is doable, but it is somewhat tedious. After all, the right-hand side
+-- of each case alternative is exactly the same! This only becomes more tedious
+-- when you deal with data types with lots of lots of data constructors. For
+-- this reason, @singletons-th@ offers the 'cases' function to generate this
+-- boilerplate code for you. The following is equivalent to the implementation
+-- of @bar@ above:
+--
+-- @
+-- bar :: SBool b -> STuple0 (Foo b)
+-- bar b = $(cases ''SBool [| b |] [| STuple0 |])
+-- @
 cases :: DsMonad q
-      => Name        -- ^ The head of the type of the scrutinee. (Like @''Maybe@ or @''Bool@.)
+      => Name        -- ^ The head of the type of the scrutinee. (Like @''Maybe@, @''Bool@, @''SBool@, etc.)
```

Comment:
Would the head of the scrutinee ever *not* be a GADTy type? Should the first two examples be removed?

Reply:
Yes, a fair point. Although nothing in the implementation of `cases` precludes the possibility of matching on a non-GADT type, it's difficult to imagine that one would ever want to do so. I've removed the mentions of `Maybe` and `Bool`.